src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.\r
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using AWS.Deploy.Common;
+using AWS.Deploy.Common.IO;
+
+namespace AWS.Deploy.Orchestration.Utilities
+{
+    public interface ICloudApplicationNameGenerator
+    {
+        /// <summary>
+        /// Generates a valid candidate for <see cref="CloudApplication.Name"/> based on <paramref name="target"/>.
+        /// Name  is checked to ensure it is unique, ie <paramref name="existingApplications"/> doesn't already have it.
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// Thrown if can't generate a valid name from <paramref name="target"/>.
+        /// </exception>
+        string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications);
+
+        /// <summary>
+        /// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html
+        /// </summary>
+        bool IsValidName(string name);
+    }
+
+    public class CloudApplicationNameGenerator : ICloudApplicationNameGenerator
+    {
+        private readonly IFileManager _fileManager;
+        private readonly IDirectoryManager _directoryManager;
+        /// <summary>
+        /// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html 
+        /// </summary>
+        private readonly Regex _validatorRegex = new ("^[a-zA-Z][a-zA-Z0-9-]{2,126}$", RegexOptions.Compiled);
+
+        public CloudApplicationNameGenerator(IFileManager fileManager, IDirectoryManager directoryManager)
+        {
+            _fileManager = fileManager;
+            _directoryManager = directoryManager;
+        }
+
+        public string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications)
+        {
+            // generate recommendation
+            var recommendedPrefix = "deployment";
+
+            if (_fileManager.Exists(target.ProjectPath))
+                recommendedPrefix = Path.GetFileNameWithoutExtension(target.ProjectPath);
```

Comment:
Customer can have a directory starting with - (dash) or number which isn't accepted by CloudFormation.

Reply:
You are correct.  Down below the `recommendedPrefix` is [sanitized](https://github.com/aws/aws-dotnet-deploy/pull/169/files#diff-a0db7211bc877b00e7c00a810aa3a6a1459080bdc78592fed5541b10b404759bR66) to remove invalid characters at the start of the recommendation.


Here's the test case covering it: https://github.com/aws/aws-dotnet-deploy/pull/169/files#diff-a0db7211bc877b00e7c00a810aa3a6a1459080bdc78592fed5541b10b404759bR66